{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "title: Notebook S9\n",
    "author: Enes Kemal Ergin\n",
    "date: \"{{ datetime.now().strftime('%Y-%m-%d %H:%M:%S') }}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Backbones of the Analysis\n",
    "\n",
    "- **License:** [MIT License](https://opensource.org/licenses/MIT)\n",
    "- **Version:** 0.2\n",
    "- **Edit Log:** \n",
    "    - 2024-01-19: Initial version of the notebook\n",
    "    - 2024-02-21: Revised the notes \n",
    "\n",
    "This notebook presents the essential packages that facilitated the data analysis and construction of the QuEStVar framework. While these packages are mentioned in the manuscript and throughout the notebooks, they are displayed here with their references, links, and the versions used.\n",
    "\n",
    "The references were primarily sourced from citebay.com, and the links direct to the project's main page. Please note that the versions used in the analysis may reflect something other than the most current version of the project at the time of writing this notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Details on the Libraries\n",
    "\n",
    "The libraries are listed in order of their criticality to the analysis. Some packages, only used once or twice, are included for completeness, but their exclusion would not disrupt the analysis.\n",
    "\n",
    "1. [Python 3](https://www.python.org/)\n",
    "    - **Version:** 3.9.18\n",
    "    - **Reference:** [Van Rossum and Drake 2009](https://dl.acm.org/doi/book/10.5555/1593511)\n",
    "    - **Description:** Python is an interpreted, high-level and general-purpose programming language.\n",
    "    - **How it was used:** The whole analysis was done in Python 3. The main libraries that were used are listed below.\n",
    "2. [Jupyter](https://jupyter.org/)\n",
    "    - **Version:** 3.5.3\n",
    "    - **Reference:** [Kluyver et al. 2016](https://ebooks.iospress.nl/publication/42900)\n",
    "    - **Description:** Jupyter is a project that supports interactive data science and scientific computing across many programming languages. It provides a browser-based notebook interface that can be used to write and execute code, and makes these forms of computing intuitive and reproducible.\n",
    "    - **How it was used:** The whole analysis was done in Jupyter notebooks. The `IPython`, `jupyterlab`, `notebook`, and `nbconvert` packages were used to create the notebooks and convert them to HTML and PDF formats.\n",
    "2. [Pandas](https://pandas.pydata.org/)\n",
    "    - **Version:** 1.5.2\n",
    "    - **Reference:** [McKinney 2010](https://conference.scipy.org/proceedings/scipy2010/mckinney.html)\n",
    "    - **Description:** Pandas is a versitile data analysis and manipulation library for Python.\n",
    "    - **How it was used:** Pandas can be found everywhere in the analysis from read/write to manipulate tables for summary, plotting, and analysis.\n",
    "3. [NumPy](https://numpy.org/)\n",
    "    - **Version:** 1.24.3\n",
    "    - **Reference:** [Harris et al. 2020](https://www.nature.com/articles/s41586-020-2649-2)\n",
    "    - **Description:** Fundamental numerical computations with efficient array operations.\n",
    "    - **How it was used:** NumPy is used to improve the complex and intense computations necessary for masked arrays and vectorized operations. Where pandas is built on top of NumPy, it is not as efficient as NumPy for these operations.\n",
    "4. [Matplotlib](https://matplotlib.org/)\n",
    "    - **Version:** 3.7.2\n",
    "    - **Reference:** [Hunter 2007](https://ieeexplore.ieee.org/document/4160265)\n",
    "    - **Description:** Matplotlib is a comprehensive plotting library for Python.\n",
    "    - **How it was used:** Matplotlib's backbone functionality is used to complement the pandas, seaborn, and other plottin specific libraries.\n",
    "5. [Seaborn](https://seaborn.pydata.org/)\n",
    "    - **Version:** 0.11.2\n",
    "    - **Reference:** [Waskom 2021](https://joss.theoj.org/papers/10.21105/joss.03021)\n",
    "    - **Description:** Seaborn is high-level interface for statistical data visualization built on top of Matplotlib.\n",
    "    - **How it was used:** Seaborn is used to create the majority of the plots along with Matplotlib.\n",
    "6. [SciPy](https://www.scipy.org/)\n",
    "    - **Version:** 1.10.1\n",
    "    - **Reference:** [Virtanen et al. 2020](https://www.nature.com/articles/s41592-019-0686-2)\n",
    "    - **Description:** SciPy is a collection of mathematical algorithms and convenience functions built on the Numpy extension of Python.\n",
    "    - **How it was used:** SciPy is used its masked-array based testing fundamental functions such as interpolate, special, and stats.\n",
    "7. [Feather-format](https://arrow.apache.org/docs/python/feather.html)\n",
    "    - **Version:** 0.4.1\n",
    "    - **Reference:** [Apache-Feather](https://github.com/wesm/feather)\n",
    "    - **Description:** Feather is a fast, lightweight, and easy-to-use binary file format for storing data frames.\n",
    "    - **How it was used:** Feather is used to store the data frames in a binary format that can be read very quickly to optimize storage and read/write time.\n",
    "8. [g:Profiler Python API](https://biit.cs.ut.ee/gprofiler/page/apis)\n",
    "    - **Version:** 1.0.0\n",
    "    - **Reference:** [Reimand et al. 2016](https://academic.oup.com/nar/article/44/W1/W83/2499353)\n",
    "    - **Description:** g:Profiler is a public web server for characterizing and manipulating gene lists.\n",
    "    - **How it was used:** g:Profiler is used to annotate the protein sets in single or multi-query modes.\n",
    "9. [BioPython](https://biopython.org/)\n",
    "    - **Version:** 1.81\n",
    "    - **Reference:** [Cock et al. 2009](https://academic.oup.com/bioinformatics/article/25/11/1422/330687)\n",
    "    - **Description:** Collection of modules for biological computation.\n",
    "    - **How it was used:** Biopython is used to parse fasta with molecular weight. Fasta to table\n",
    "10. [Polars](https://docs.pola.rs/user-guide/)\n",
    "    - **Version:** 0.18.1\n",
    "    - **Reference:** [Pola.rs](https://github.com/pola-rs/polars)\n",
    "    - **Description:** Rust-based very fast implementation of data handling with labeled 2d tables. It is a new-generation and rapidly evolving library, and likely will replace Pandas in the future. \n",
    "    - **How it was used:** Polars is used for very intense calculations that are best to be done in labeled table setting such as protein_status_matrix. \n",
    "11. [Sklearn](https://scikit-learn.org/stable/)\n",
    "    - **Version:** 1.1.1\n",
    "    - **Reference:** [Pedregosa et al. 2011](https://www.semanticscholar.org/paper/Scikit-learn%3A-Machine-Learning-in-Python-Pedregosa-Varoquaux/168f28ac3c8c7ea63bf7ed25f2288e8b67e2fe74)\n",
    "    - **Description:** Collection of machine learning functionality\n",
    "    - **How it was used:** Used for dimensional reduction methods such as PCA and tSNE and clustering methods such as KMeans and DBSCAN, when needed.\n",
    "12. [PyComplexHeatmap](https://dingwb.github.io/PyComplexHeatmap/build/html/index.html)\n",
    "    - **Version:** 1.6.4\n",
    "    - **Reference:** [Ding et al. 2023](https://onlinelibrary.wiley.com/doi/10.1002/imt2.115)\n",
    "    - **Description:** Plotting library for complex heatmaps\n",
    "    - **How it was used:** Used to add enhanced features to seaborns clustermap.\n",
    "13. [Upsetplot](https://upsetplot.readthedocs.io/en/stable/)\n",
    "    - **Version:** 0.7.0\n",
    "    <!-- - **Reference:** [Ding et al. 2023](https://onlinelibrary.wiley.com/doi/10.1002/imt2.115) -->\n",
    "    - **Description:** Plotting upset plots\n",
    "    - **How it was used:** Upset plots are used to display the overlap of the protein sets from more than 3 sets in the analysis.\n",
    "14. [session_info](https://gitlab.com/joelostblom/session_info)\n",
    "    - **Version:** 1.0.0\n",
    "    <!-- - **Reference:** [Ding et al. 2023](https://onlinelibrary.wiley.com/doi/10.1002/imt2.115) -->\n",
    "    - **Description:** Simple function to display the session info of the libraries used in the analysis.\n",
    "    - **How it was used:** As it is intended to be used.\n",
    "\n",
    "\n",
    "## References\n",
    "\n",
    "1. Van Rossum, G., & Drake, F. L. (2009). Python 3 Reference Manual. Scotts Valley, CA: CreateSpace.\n",
    "1. Kluyver, T., Ragan-Kelley, B., Fernando P&#x27;erez, Granger, B., Bussonnier, M., Frederic, J., … Willing, C. (2016). Jupyter Notebooks – a publishing format for reproducible computational workflows. In F. Loizides & B. Schmidt (Eds.), Positioning and Power in Academic Publishing: Players, Agents and Agendas (pp. 87–90). IOS Press. https://doi.org/10.3233/978-1-61499-649-1-87\n",
    "1. McKinney, W. (2010). Data structures for statistical computing in python. In Proceedings of the 9th Python in Science Conference (Vol. 445, pp. 51–56). Austin, TX.\n",
    "1. Harris, C. R., Millman, K. J., van der Walt, S. J., Gommers, R., Virtanen, P., Cournapeau, D., … Oliphant, T. E. (2020). Array programming with NumPy. Nature, 585(7825), 357–362. https://doi.org/10.1038/s41586-020-2649-2\n",
    "1. Hunter, J. D. (2007). Matplotlib: A 2D graphics environment. Computing in Science and Engineering, 9(3), 90–95. https://doi.org/10.1109/MCSE.2007.55\n",
    "1. Waskom, M. L., (2021). seaborn: statistical data visualization. Journal of Open Source Software, 6(60), 3021, https://doi.org/10.21105/joss.03021\n",
    "1. Virtanen, P., Gommers, R., Oliphant, T. E., Haberland, M., Reddy, T., Cournapeau, D., … SciPy 1.0 Contributors. (2020). SciPy 1.0: Fundamental Algorithms for Scientific Computing in Python. Nature Methods, 17, 261–272. https://doi.org/10.1038/s41592-019-0686-2\n",
    "1. Feather File Format — Apache Arrow v15.0.0, https://arrow.apache.org/docs/python/feather.html (accessed January 23, 2024).\n",
    "1. Reimand, J., Arak, T., Adler, P., Kolberg, L., Reisberg, S., Peterson, H., & Vilo, J. (2016). g:Profiler—a web server for functional interpretation of gene lists (2016 update). Nucleic Acids Research, 44(W1), W83–W89. https://doi.org/10.1093/nar/gkw199\n",
    "1. Cock, P. J., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., … others. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics, 25(11), 1422–1423.\n",
    "1. pola-rs/polars: Dataframes powered by a multithreaded, vectorized query engine, written in Rust, https://github.com/pola-rs/polars (accessed January 23, 2024).\n",
    "1. Pedregosa, F., Varoquaux, Ga\"el, Gramfort, A., Michel, V., Thirion, B., Grisel, O., … others. (2011). Scikit-learn: Machine learning in Python. Journal of Machine Learning Research, 12(Oct), 2825–2830.\n",
    "1. Ding, W., Goldberg, D. and Zhou, W. (2023), PyComplexHeatmap: A Python package to visualize multimodal genomics data. iMeta e115. https://doi.org/10.1002/imt2.115\n",
    "\n",
    "\n",
    "## The Session Info\n",
    "\n",
    "The session info is used to show the full detail of the used packages, their dependencies, and the environment the analysis is conducted at. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "Bio                 1.81\n",
       "PyComplexHeatmap    1.6.4\n",
       "feather             0.4.1\n",
       "gprofiler           1.0.0\n",
       "matplotlib          3.7.2\n",
       "numpy               1.24.3\n",
       "pandas              1.5.2\n",
       "polars              0.18.1\n",
       "scipy               1.10.1\n",
       "seaborn             0.11.2\n",
       "session_info        1.0.0\n",
       "sklearn             1.1.1\n",
       "upsetplot           0.7.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                 9.3.0\n",
       "asttokens           NA\n",
       "backcall            0.2.0\n",
       "beta_ufunc          NA\n",
       "binom_ufunc         NA\n",
       "bottleneck          1.3.7\n",
       "brotli              1.0.9\n",
       "certifi             2023.11.17\n",
       "cffi                1.16.0\n",
       "chardet             5.2.0\n",
       "charset_normalizer  3.3.0\n",
       "cloudpickle         2.2.1\n",
       "colorama            0.4.6\n",
       "comm                0.1.4\n",
       "cycler              0.12.1\n",
       "cython_runtime      NA\n",
       "dateutil            2.8.2\n",
       "debugpy             1.8.0\n",
       "decorator           5.1.1\n",
       "defusedxml          0.7.1\n",
       "django              3.2.16\n",
       "entrypoints         0.4\n",
       "exceptiongroup      1.1.3\n",
       "executing           1.2.0\n",
       "google              NA\n",
       "hypergeom_ufunc     NA\n",
       "idna                3.4\n",
       "importlib_metadata  NA\n",
       "importlib_resources NA\n",
       "invgauss_ufunc      NA\n",
       "ipykernel           6.25.2\n",
       "ipython_genutils    0.2.0\n",
       "ipywidgets          8.1.1\n",
       "jedi                0.19.1\n",
       "joblib              1.3.2\n",
       "jupyter_server      2.7.3\n",
       "kiwisolver          1.4.5\n",
       "mkl                 2.4.0\n",
       "mpl_toolkits        NA\n",
       "nbinom_ufunc        NA\n",
       "ncf_ufunc           NA\n",
       "nct_ufunc           NA\n",
       "ncx2_ufunc          NA\n",
       "numexpr             2.8.7\n",
       "packaging           23.2\n",
       "parso               0.8.3\n",
       "patsy               0.5.5\n",
       "pexpect             4.8.0\n",
       "pickleshare         0.7.5\n",
       "platformdirs        3.11.0\n",
       "prompt_toolkit      3.0.39\n",
       "psutil              5.9.5\n",
       "ptyprocess          0.7.0\n",
       "pure_eval           0.2.2\n",
       "pyarrow             11.0.0\n",
       "pydev_ipython       NA\n",
       "pydevconsole        NA\n",
       "pydevd              2.9.5\n",
       "pydevd_file_utils   NA\n",
       "pydevd_plugins      NA\n",
       "pydevd_tracing      NA\n",
       "pygments            2.16.1\n",
       "pyparsing           3.0.9\n",
       "pytz                2023.3.post1\n",
       "requests            2.31.0\n",
       "setuptools          68.2.2\n",
       "six                 1.16.0\n",
       "skewnorm_ufunc      NA\n",
       "socks               1.7.1\n",
       "sphinxcontrib       NA\n",
       "stack_data          0.6.2\n",
       "statsmodels         0.14.1\n",
       "threadpoolctl       3.2.0\n",
       "tornado             6.2\n",
       "traitlets           5.11.2\n",
       "typing_extensions   NA\n",
       "urllib3             2.0.6\n",
       "wcwidth             0.2.13\n",
       "zipp                NA\n",
       "zmq                 24.0.1\n",
       "zoneinfo            NA\n",
       "zope                NA\n",
       "zstandard           0.21.0\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             8.16.1\n",
       "jupyter_client      7.4.9\n",
       "jupyter_core        5.3.2\n",
       "jupyterlab          3.5.3\n",
       "notebook            6.5.6\n",
       "-----\n",
       "Python 3.9.18 (main, Sep 11 2023, 13:41:44) [GCC 11.2.0]\n",
       "Linux-6.6.6-76060606-generic-x86_64-with-glibc2.35\n",
       "-----\n",
       "Session information updated at 2024-01-27 15:25\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import session_info\n",
    "\n",
    "import feather as ft\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "\n",
    "import scipy as sp\n",
    "import sklearn as sk\n",
    "import Bio as bio\n",
    "import gprofiler as gp\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib as plt\n",
    "import upsetplot as up\n",
    "import PyComplexHeatmap as pch\n",
    "\n",
    "session_info.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
