{
 "cells": [
  {
   "cell_type": "raw",
   "id": "8bb27eca",
   "metadata": {},
   "source": [
    "title: Notebook S4\n",
    "author: Enes Kemal Ergin\n",
    "date: \"{{ datetime.now().strftime('%Y-%m-%d %H:%M:%S') }}\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e2a12c91-b36f-48a7-bec4-0f556a232cbe",
   "metadata": {},
   "source": [
    "# Prepare Cell Line Metadata \n",
    "\n",
    "- **License:** [MIT License](https://opensource.org/licenses/MIT)\n",
    "- **Version:** 0.2\n",
    "- **Edit Log:** \n",
    "    - 2024-01-19: Initial version of the notebook\n",
    "    - 2024-02-21: Revised the notes \n",
    "\n",
    "\n",
    "**Notebook Summary:** \n",
    "\n",
    "In this notebook, I will prepare the cell line metadata with enhancements and additional information to analyze this dataset. I will also create smaller helper data here, such as Reactome pathway hierarchy reference and detailed protein info data from UniProt.  \n",
    "\n",
    "To expand the cell line information and create comprehensive metadata, I've done the following steps:\n",
    "1. Grab the model annotation data from the [DepMap Models](https://depmap.org/portal/download/)\n",
    "2. Merge with the model annotation and add the following information:\n",
    "    - sample site\n",
    "    - cancer type detail\n",
    "    - tissue status\n",
    "    - age \n",
    "    - Sex \n",
    "3. Due to some of those information is left unknown create a manual curation to fill those especially the age.\n",
    "4. Adding a higher variable called \"System Name\" indicating the system that the cell line is originated from. The Tissue to System mapper is created from the [\"National Cancer Institute\"(NCI)](https://www.cancer.gov/types/by-body-location)\n",
    "5. Also removed cell lines that have missing annotation in Age, Tumour Status (Primary, Metastatic), Sex, etc. \n",
    "6. Removed some cell lines from the lymphoid system since there are a lot of diversity in the subtypes which might skew the results in the favour of lymphoid system.\n",
    "\n",
    "---\n",
    "\n",
    "**Data Information:**\n",
    "\n",
    "The data used in this notebook and this part of the analysis is from [GonÃ§alves et al. 2022](https://www.cell.com/cancer-cell/fulltext/S1535-6108(22)00274-4) and the data can be downloaded from [figshare Repo](https://figshare.com/articles/dataset/Pan-cancer_proteomic_map_of_949_human_cell_lines/19345397). For this notebook I am going to be using the following files accessed on 2023-11-01: \n",
    "- `ProCan-DepMapSanger_mapping_file_replicates.txt` - Containing a metadata of the samples at a replicate level, I've saved this as `ProCan-DepMapSanger_Metadata.txt` for my own convenience in the `./data/raw/` folder. \n",
    "\n",
    "Additionally I am also going to be using the following data from the [DepMap Portal](https://depmap.org/portal/download/) accessed on 2023-11-01:\n",
    "- `Model_v2.csv` - Containing a cell line model information in depth, I've saved this as `DepMap_Model_Annotations.csv` for my own convenience in the `./data/raw/` folder.\n",
    "\n",
    "For the manual additions and curations I've made the following files and place them in the `./data/raw/` folder:\n",
    "- `Tissue_System_Map.csv` - Contains the mapping of tissues indicated in the depMap data to the systems for higher level grouping.\n",
    "- `manuscript_cell_line_annotations-manual.csv` - This is the manually updated annotations for the cell lines that have missing information in the depMap data.\n",
    "\n",
    "---\n",
    "\n",
    "For the reactome pathway hierarchy helper dataset I will be using the following data from the [Reactome](https://reactome.org/download-data) accessed on 2023-11-01:\n",
    "- `ReactomePathways.txt` - Containing all pathways available in the database\n",
    "- `ReactomePathwaysRelation.txt` - Containing a pathway hierarchy relationships eg. col1 child, col2 parent\n",
    "- Both of them saved as is in the `./data/raw/` folder.\n",
    "\n",
    "---\n",
    "\n",
    "For the fasta mapper, I've downloaded human \"reviewed proteome without isoforms\" fasta from [Uniprot](https://www.uniprot.org/) accessed on 2023-11-01 and saved it in the `./data/raw/` folder as `HUMAN_REVIEWED_2023_11_01.fasta`. \n",
    "\n",
    "\n",
    "## Setup Notebook\n",
    "\n",
    "This part is a standard for my notebooks, where I import the all used libraries, set the design of the notebook, and define the paths for data and figures. \n",
    "\n",
    "> **Note:** The HTML rendering of this notebook will not show the code cells by default, but the code can be visible by clicking the \"code\" buttons aligned on the right.\n",
    "\n",
    "### Libraries Used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8a13433-0aa8-44e3-b5c6-b7277e2266a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import warnings\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "sys.path.append('../')          # Expand the path to parent dir\n",
    "from questvar import utils      # Import custom modules\n",
    " \n",
    "warnings.filterwarnings('ignore')\n",
    "# Initialize the timer\n",
    "startTime = utils.getTime()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de2bf87c",
   "metadata": {},
   "source": [
    "### Notebook's Design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0e26d004",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Configure dataframe displaying\n",
    "pd.set_option('display.float_format', lambda x: '%.4f' % x)\n",
    "pd.set_option('display.max_rows', 25)\n",
    "pd.set_option('display.max_columns', 25)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4672e900-811b-4e54-91e8-28fec389adc0",
   "metadata": {},
   "source": [
    "### Data Paths\n",
    "\n",
    "I keep the data and figures folders seperate to allow for a more organized structure. Also within the data folder I create raw, processed, and in later notebook results folder to seperate each step of the process. \n",
    "\n",
    "The raw `data/raw/` folder needs to be present since it should contain the original data. For only this notebook I will save the expanded metadata and other helper data in the `data/raw/` folder as well since I consider this notebook to be pre-organizational step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1164b92e-bb9b-477a-ab32-5e7881c7f777",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Establish the paths\n",
    "data_path = \"./data/\"\n",
    "input_path = f\"{data_path}raw/\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2943d531-5474-41ab-856e-27cb1fda1bda",
   "metadata": {},
   "source": [
    "# Expand the Existing Metadata\n",
    "\n",
    "The metadata and cell line information in the manuscript's data folder didn't have certain atributes and had some missing information which prompted to do some expansion with depMap model informations and manual curations. "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c2861437-40e7-4334-b626-8bb250914700",
   "metadata": {},
   "source": [
    "## Sample Metadata from the Manuscript\n",
    "\n",
    "This is the replicate level metadata from the manuscript, containing the \"Automatic_MS_filename\" as the column names, and other information such as batch info, cell line id, tissue, cancer, and subtype infos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a4b2e7fe-1e03-4c7e-97f7-650069d29892",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data shape: (6864, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Automatic_MS_filename</th>\n",
       "      <th>Batch</th>\n",
       "      <th>Date</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>Cell_line</th>\n",
       "      <th>SIDM</th>\n",
       "      <th>Project_Identifier</th>\n",
       "      <th>Tissue_type</th>\n",
       "      <th>Cancer_type</th>\n",
       "      <th>Cancer_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>191005_b43-t1-13_00dlf_00jcq_m01_s_1</td>\n",
       "      <td>P04</td>\n",
       "      <td>191005</td>\n",
       "      <td>M01</td>\n",
       "      <td>201T</td>\n",
       "      <td>SIDM00055</td>\n",
       "      <td>SIDM00055;201T</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Non-Small Cell Lung Carcinoma</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>191006_b45-t1-13_00dmb_00jd7_m01_s_1</td>\n",
       "      <td>P04</td>\n",
       "      <td>191006</td>\n",
       "      <td>M01</td>\n",
       "      <td>201T</td>\n",
       "      <td>SIDM00055</td>\n",
       "      <td>SIDM00055;201T</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Non-Small Cell Lung Carcinoma</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>191008_b47-t1-13_00dn7_00jfo_m01_s_1</td>\n",
       "      <td>P04</td>\n",
       "      <td>191008</td>\n",
       "      <td>M01</td>\n",
       "      <td>201T</td>\n",
       "      <td>SIDM00055</td>\n",
       "      <td>SIDM00055;201T</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Non-Small Cell Lung Carcinoma</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>191008_b45-t2-13_00dmb_00jep_m03_s_1</td>\n",
       "      <td>P04</td>\n",
       "      <td>191008</td>\n",
       "      <td>M03</td>\n",
       "      <td>201T</td>\n",
       "      <td>SIDM00055</td>\n",
       "      <td>SIDM00055;201T</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Non-Small Cell Lung Carcinoma</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>191009_b43-t2-13_00dlf_00jg8_m03_s_1</td>\n",
       "      <td>P04</td>\n",
       "      <td>191009</td>\n",
       "      <td>M03</td>\n",
       "      <td>201T</td>\n",
       "      <td>SIDM00055</td>\n",
       "      <td>SIDM00055;201T</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Non-Small Cell Lung Carcinoma</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Automatic_MS_filename Batch    Date Instrument Cell_line  \\\n",
       "0  191005_b43-t1-13_00dlf_00jcq_m01_s_1   P04  191005        M01      201T   \n",
       "1  191006_b45-t1-13_00dmb_00jd7_m01_s_1   P04  191006        M01      201T   \n",
       "2  191008_b47-t1-13_00dn7_00jfo_m01_s_1   P04  191008        M01      201T   \n",
       "3  191008_b45-t2-13_00dmb_00jep_m03_s_1   P04  191008        M03      201T   \n",
       "4  191009_b43-t2-13_00dlf_00jg8_m03_s_1   P04  191009        M03      201T   \n",
       "\n",
       "        SIDM Project_Identifier Tissue_type                    Cancer_type  \\\n",
       "0  SIDM00055     SIDM00055;201T        Lung  Non-Small Cell Lung Carcinoma   \n",
       "1  SIDM00055     SIDM00055;201T        Lung  Non-Small Cell Lung Carcinoma   \n",
       "2  SIDM00055     SIDM00055;201T        Lung  Non-Small Cell Lung Carcinoma   \n",
       "3  SIDM00055     SIDM00055;201T        Lung  Non-Small Cell Lung Carcinoma   \n",
       "4  SIDM00055     SIDM00055;201T        Lung  Non-Small Cell Lung Carcinoma   \n",
       "\n",
       "        Cancer_subtype  \n",
       "0  Lung Adenocarcinoma  \n",
       "1  Lung Adenocarcinoma  \n",
       "2  Lung Adenocarcinoma  \n",
       "3  Lung Adenocarcinoma  \n",
       "4  Lung Adenocarcinoma  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replica_meta = pd.read_csv(\n",
    "    input_path+'ProCan-DepMapSanger_Metadata.txt', sep=\"\\t\"\n",
    ").sort_values([\n",
    "    \"Cell_line\", \"Batch\", \"Instrument\", \"Date\"\n",
    "]).reset_index(drop=True)\n",
    "replica_meta[\"Date\"] = replica_meta[\"Automatic_MS_filename\"].str.split('_').str[0]\n",
    "print(\"Data shape:\", replica_meta.shape)\n",
    "replica_meta.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "86c97a01-0668-4d37-aaa0-a5d67decc55f",
   "metadata": {},
   "source": [
    "## Updating the Metadata with Simple Navigation Columns\n",
    "\n",
    "I add the following columns to the metadata to allow for easier navigation and filtering in the future.\n",
    "\n",
    "- **VarianceID** - This is the unique batch identifier put together by joining \"Instrument\" and \"Batch\" columns.\n",
    "- **Replicate** - Replicate number indicator per VarianceID\n",
    "- **UniqueSample** - Unique sample identifier per cell line, this is created by joining the \"VarianceID\" and \"SIDM\" columns. (SIDM is the cell line model id)\n",
    "- **ColumnName** - New simpler column names used by combining \"UniqueSample\" and \"Replicate\" columns. This is used to make the column names shorter and easier to read.\n",
    "\n",
    "Then I select the columns that I want to keep; `[\"Automatic_MS_filename\", \"ColumnName\", \"UniqueSample\", \"VarianceID\", \"Cell_line\", \"Replicate\", \"SIDM\"]`\n",
    "\n",
    "Finally I drop the `Control` samples from the cell lines since I want to focus on only the cancer cell lines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6970cdcb-748c-4307-b4ec-806f7ee18a71",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data shape: (5800, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Automatic_MS_filename</th>\n",
       "      <th>ColumnName</th>\n",
       "      <th>UniqueSample</th>\n",
       "      <th>VarianceID</th>\n",
       "      <th>Cell_line</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>SIDM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>191005_b43-t1-13_00dlf_00jcq_m01_s_1</td>\n",
       "      <td>M01|P04|SIDM00055|1</td>\n",
       "      <td>M01|P04|SIDM00055</td>\n",
       "      <td>M01|P04</td>\n",
       "      <td>201T</td>\n",
       "      <td>1</td>\n",
       "      <td>SIDM00055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>191006_b45-t1-13_00dmb_00jd7_m01_s_1</td>\n",
       "      <td>M01|P04|SIDM00055|2</td>\n",
       "      <td>M01|P04|SIDM00055</td>\n",
       "      <td>M01|P04</td>\n",
       "      <td>201T</td>\n",
       "      <td>2</td>\n",
       "      <td>SIDM00055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>191008_b47-t1-13_00dn7_00jfo_m01_s_1</td>\n",
       "      <td>M01|P04|SIDM00055|3</td>\n",
       "      <td>M01|P04|SIDM00055</td>\n",
       "      <td>M01|P04</td>\n",
       "      <td>201T</td>\n",
       "      <td>3</td>\n",
       "      <td>SIDM00055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>191008_b45-t2-13_00dmb_00jep_m03_s_1</td>\n",
       "      <td>M03|P04|SIDM00055|1</td>\n",
       "      <td>M03|P04|SIDM00055</td>\n",
       "      <td>M03|P04</td>\n",
       "      <td>201T</td>\n",
       "      <td>1</td>\n",
       "      <td>SIDM00055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>191009_b43-t2-13_00dlf_00jg8_m03_s_1</td>\n",
       "      <td>M03|P04|SIDM00055|2</td>\n",
       "      <td>M03|P04|SIDM00055</td>\n",
       "      <td>M03|P04</td>\n",
       "      <td>201T</td>\n",
       "      <td>2</td>\n",
       "      <td>SIDM00055</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Automatic_MS_filename           ColumnName  \\\n",
       "0  191005_b43-t1-13_00dlf_00jcq_m01_s_1  M01|P04|SIDM00055|1   \n",
       "1  191006_b45-t1-13_00dmb_00jd7_m01_s_1  M01|P04|SIDM00055|2   \n",
       "2  191008_b47-t1-13_00dn7_00jfo_m01_s_1  M01|P04|SIDM00055|3   \n",
       "3  191008_b45-t2-13_00dmb_00jep_m03_s_1  M03|P04|SIDM00055|1   \n",
       "4  191009_b43-t2-13_00dlf_00jg8_m03_s_1  M03|P04|SIDM00055|2   \n",
       "\n",
       "        UniqueSample VarianceID Cell_line Replicate       SIDM  \n",
       "0  M01|P04|SIDM00055    M01|P04      201T         1  SIDM00055  \n",
       "1  M01|P04|SIDM00055    M01|P04      201T         2  SIDM00055  \n",
       "2  M01|P04|SIDM00055    M01|P04      201T         3  SIDM00055  \n",
       "3  M03|P04|SIDM00055    M03|P04      201T         1  SIDM00055  \n",
       "4  M03|P04|SIDM00055    M03|P04      201T         2  SIDM00055  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create variance ID to hold Instrument and Batch info as one\n",
    "replica_meta[\"VarianceID\"] = replica_meta.apply(\n",
    "    lambda x: \"|\".join([x[\"Instrument\"], x[\"Batch\"]]), \n",
    "    axis=1\n",
    ")\n",
    "# Create Replica info for each sample belong to same cell line - variance ID combo\n",
    "replica_meta[\"Replicate\"] = \"1\"\n",
    "for n, g in replica_meta.groupby([\"VarianceID\", \"SIDM\"]):\n",
    "    if g.shape[0] > 1:\n",
    "        replica_meta.loc[\n",
    "            (replica_meta[\"VarianceID\"]==n[0]) & \n",
    "            (replica_meta[\"SIDM\"]==n[1]), \n",
    "            \"Replicate\"\n",
    "        ] = [str(i+1) for i in range(g.shape[0])]\n",
    "# Create UniqueSample column indicating cell line - variance id\n",
    "replica_meta[\"UniqueSample\"] = replica_meta.apply(\n",
    "    lambda x: \"|\".join([x[\"VarianceID\"], x[\"SIDM\"]]), \n",
    "    axis=1\n",
    ")\n",
    "# Create Unique column name using uniqueSample and replicate\n",
    "replica_meta[\"ColumnName\"] = replica_meta.apply(\n",
    "    lambda x: \"|\".join([x[\"UniqueSample\"], x[\"Replicate\"]]), \n",
    "    axis=1\n",
    ")\n",
    "# Remove columns that are not gonna be used\n",
    "replica_meta = replica_meta[[\n",
    "    \"Automatic_MS_filename\", \"ColumnName\", \"UniqueSample\", \n",
    "    \"VarianceID\", \"Cell_line\", \"Replicate\", \"SIDM\"\n",
    "]]\n",
    "\n",
    "# Remove the Control_HEK293T samples\n",
    "replica_meta = replica_meta[replica_meta[\"SIDM\"] != 'Control_HEK293T']\n",
    "# Preview the results\n",
    "print(\"Data shape:\", replica_meta.shape)\n",
    "replica_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb41812c",
   "metadata": {},
   "source": [
    "## Open and Clean DepMap Models Annotations\n",
    "\n",
    "DepMap Models annotation data contains a lot of information about the cell lines, but I am only interested in the following columns:\n",
    "- **SangerModelID** - Cell line model id (used to join with the manuscript's metadata)\n",
    "- **CellLineName** - Cell line name (used to join with the manuscript's metadata)\n",
    "- **AgeCategory** - Age category of the cell line\n",
    "- **PrimaryOrMetastasis** - Tumour status of the cell line\n",
    "- **Sex** - The sex of the patient that the cell line is derived from\n",
    "- **SampleCollectionSite** - The site where the sample is collected from\n",
    "- **OncotreeLineage** - The lineage of the cell line\n",
    "- **OncotreePrimaryDisease** - The primary disease of the cell line\n",
    "- **OncotreeSubtype** - The subtype of the cell line\n",
    "\n",
    "I am particularly interested in the `AgeCategory`, `PrimaryOrMetastasis`, `Sex`, `SampleCollectionSite`, `OncotreeLineage`, `OncotreePrimaryDisease`, and `OncotreeSubtype` columns, which can be used in various groupings when testing and summarizing the data.\n",
    "\n",
    "Using DepMap Models annotation data also changes the tissue names for the cell line origins for certain cancers for instance, originally myleoid and lymphoid cancers are all grouped under the Hematopoietic and Lymphoid Tissue but with the DepMap Models annotation data I can separate them into Myeloid and Lymphoid systems. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8da2ab53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Models in DepMap: 1864\n",
      "Unmatched cell line ids: {'SIDM01117', 'SIDM00427'}\n",
      "Number of Cell Line info in DepMap: 950\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AgeCategory</th>\n",
       "      <th>PrimaryOrMetastasis</th>\n",
       "      <th>Sex</th>\n",
       "      <th>SampleCollectionSite</th>\n",
       "      <th>OncotreeLineage</th>\n",
       "      <th>OncotreePrimaryDisease</th>\n",
       "      <th>OncotreeSubtype</th>\n",
       "      <th>Cell_line</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SIDM00018</th>\n",
       "      <td>Adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "      <td>K052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00023</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>oesophagus</td>\n",
       "      <td>Esophagus/Stomach</td>\n",
       "      <td>Esophageal Squamous Cell Carcinoma</td>\n",
       "      <td>Esophageal Squamous Cell Carcinoma</td>\n",
       "      <td>TE-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00040</th>\n",
       "      <td>Adult</td>\n",
       "      <td>Metastatic</td>\n",
       "      <td>Male</td>\n",
       "      <td>lymph_node</td>\n",
       "      <td>Esophagus/Stomach</td>\n",
       "      <td>Esophagogastric Adenocarcinoma</td>\n",
       "      <td>Stomach Adenocarcinoma</td>\n",
       "      <td>TMK-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00041</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>soft_tissue</td>\n",
       "      <td>Esophagus/Stomach</td>\n",
       "      <td>Esophagogastric Adenocarcinoma</td>\n",
       "      <td>Stomach Adenocarcinoma</td>\n",
       "      <td>STS-0421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00042</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Pancreatic Adenocarcinoma</td>\n",
       "      <td>Pancreatic Adenocarcinoma</td>\n",
       "      <td>PL4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          AgeCategory PrimaryOrMetastasis      Sex SampleCollectionSite  \\\n",
       "SIDM00018       Adult                 NaN     Male          bone_marrow   \n",
       "SIDM00023     Unknown                 NaN  Unknown           oesophagus   \n",
       "SIDM00040       Adult          Metastatic     Male           lymph_node   \n",
       "SIDM00041     Unknown                 NaN  Unknown          soft_tissue   \n",
       "SIDM00042     Unknown             Primary  Unknown             pancreas   \n",
       "\n",
       "             OncotreeLineage              OncotreePrimaryDisease  \\\n",
       "SIDM00018            Myeloid              Acute Myeloid Leukemia   \n",
       "SIDM00023  Esophagus/Stomach  Esophageal Squamous Cell Carcinoma   \n",
       "SIDM00040  Esophagus/Stomach      Esophagogastric Adenocarcinoma   \n",
       "SIDM00041  Esophagus/Stomach      Esophagogastric Adenocarcinoma   \n",
       "SIDM00042           Pancreas           Pancreatic Adenocarcinoma   \n",
       "\n",
       "                              OncotreeSubtype Cell_line  \n",
       "SIDM00018              Acute Myeloid Leukemia      K052  \n",
       "SIDM00023  Esophageal Squamous Cell Carcinoma     TE-12  \n",
       "SIDM00040              Stomach Adenocarcinoma     TMK-1  \n",
       "SIDM00041              Stomach Adenocarcinoma  STS-0421  \n",
       "SIDM00042           Pancreatic Adenocarcinoma       PL4  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depmap_info= pd.read_csv(\n",
    "    input_path+\"DepMap_Model_Annotations.csv\",\n",
    ")[[\n",
    "    \"SangerModelID\",\n",
    "    \"CellLineName\",\n",
    "    \"AgeCategory\",\n",
    "    \"PrimaryOrMetastasis\",\n",
    "    # \"GrowthPattern\",\n",
    "    \"Sex\",\n",
    "    \"SampleCollectionSite\",\n",
    "    # \"CCLEName\", \n",
    "    \"OncotreeLineage\",\n",
    "    \"OncotreePrimaryDisease\",\n",
    "    \"OncotreeSubtype\",\n",
    "]].set_index(\"SangerModelID\")\n",
    "print(\"Number of Models in DepMap:\", depmap_info.shape[0])\n",
    "\n",
    "# Get the Unique Cell Ids from paper\n",
    "paper_cell_ids = set(replica_meta[\"SIDM\"])\n",
    "print(\"Unmatched cell line ids:\", (paper_cell_ids - set(depmap_info.index)))\n",
    "# # Only select cell lines that are in the paper\n",
    "depmap_info = depmap_info.loc[\n",
    "    list(set.intersection(paper_cell_ids, depmap_info.index))\n",
    "]\n",
    "\n",
    "# Add unmatched cell lines to the depmap info with NA values\n",
    "depmap_info = pd.concat([\n",
    "    depmap_info,\n",
    "    pd.DataFrame(\n",
    "        index=list(paper_cell_ids - set(depmap_info.index)),\n",
    "        columns=depmap_info.columns\n",
    "    )\n",
    "])\n",
    "\n",
    "# Replace the cell line names in the depmap info with the paper's cell line names\n",
    "# Merge the cell line names from the paper to the depmap info\n",
    "depmap_info = depmap_info.merge(\n",
    "    replica_meta[[\n",
    "        \"SIDM\", \"Cell_line\"\n",
    "    ]].drop_duplicates().set_index(\"SIDM\"),\n",
    "    left_index=True, \n",
    "    right_index=True, \n",
    "    how=\"left\"\n",
    ").drop(columns=[\"CellLineName\"])\n",
    "\n",
    "print(\"Number of Cell Line info in DepMap:\", depmap_info.shape[0])\n",
    "depmap_info.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27ac0280",
   "metadata": {},
   "source": [
    "After cleaning the data and selecting cell line models that are present in the manuscript now I have a data that contains the cell line information. However this has a lot of Unknown and NaN values, which might be updated by manual check-up and curation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "100ad3a0",
   "metadata": {},
   "source": [
    "## Adding System Names to DepMap Models Annotations\n",
    "\n",
    "One additional information that I want to add to the metadata is the system name of the cell line, which is the higher level grouping of the tissue. I've created a mapper from the tissue names in the DepMap Models annotation data to the system names using the [\"National Cancer Institute\"(NCI)](https://www.cancer.gov/types/by-body-location) as the reference. The `Tissue_System_Map.csv` file contains the mapping of the tissues indicated in the depMap data to the systems for higher level grouping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "81d98678",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Cell Line info in DepMap: 950\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>AgeCategory</th>\n",
       "      <th>PrimaryOrMetastasis</th>\n",
       "      <th>Sex</th>\n",
       "      <th>SampleCollectionSite</th>\n",
       "      <th>OncotreeLineage</th>\n",
       "      <th>OncotreePrimaryDisease</th>\n",
       "      <th>OncotreeSubtype</th>\n",
       "      <th>Cell_line</th>\n",
       "      <th>System_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM00018</td>\n",
       "      <td>Adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "      <td>Acute Myeloid Leukemia</td>\n",
       "      <td>K052</td>\n",
       "      <td>Hematologic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00023</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>oesophagus</td>\n",
       "      <td>Esophagus/Stomach</td>\n",
       "      <td>Esophageal Squamous Cell Carcinoma</td>\n",
       "      <td>Esophageal Squamous Cell Carcinoma</td>\n",
       "      <td>TE-12</td>\n",
       "      <td>Digestive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM00040</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Metastatic</td>\n",
       "      <td>Male</td>\n",
       "      <td>lymph_node</td>\n",
       "      <td>Esophagus/Stomach</td>\n",
       "      <td>Esophagogastric Adenocarcinoma</td>\n",
       "      <td>Stomach Adenocarcinoma</td>\n",
       "      <td>TMK-1</td>\n",
       "      <td>Digestive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDM00041</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>soft_tissue</td>\n",
       "      <td>Esophagus/Stomach</td>\n",
       "      <td>Esophagogastric Adenocarcinoma</td>\n",
       "      <td>Stomach Adenocarcinoma</td>\n",
       "      <td>STS-0421</td>\n",
       "      <td>Digestive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDM00042</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Pancreatic Adenocarcinoma</td>\n",
       "      <td>Pancreatic Adenocarcinoma</td>\n",
       "      <td>PL4</td>\n",
       "      <td>Digestive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       index AgeCategory PrimaryOrMetastasis      Sex SampleCollectionSite  \\\n",
       "0  SIDM00018       Adult                 NaN     Male          bone_marrow   \n",
       "1  SIDM00023     Unknown                 NaN  Unknown           oesophagus   \n",
       "2  SIDM00040       Adult          Metastatic     Male           lymph_node   \n",
       "3  SIDM00041     Unknown                 NaN  Unknown          soft_tissue   \n",
       "4  SIDM00042     Unknown             Primary  Unknown             pancreas   \n",
       "\n",
       "     OncotreeLineage              OncotreePrimaryDisease  \\\n",
       "0            Myeloid              Acute Myeloid Leukemia   \n",
       "1  Esophagus/Stomach  Esophageal Squamous Cell Carcinoma   \n",
       "2  Esophagus/Stomach      Esophagogastric Adenocarcinoma   \n",
       "3  Esophagus/Stomach      Esophagogastric Adenocarcinoma   \n",
       "4           Pancreas           Pancreatic Adenocarcinoma   \n",
       "\n",
       "                      OncotreeSubtype Cell_line  System_name  \n",
       "0              Acute Myeloid Leukemia      K052  Hematologic  \n",
       "1  Esophageal Squamous Cell Carcinoma     TE-12    Digestive  \n",
       "2              Stomach Adenocarcinoma     TMK-1    Digestive  \n",
       "3              Stomach Adenocarcinoma  STS-0421    Digestive  \n",
       "4           Pancreatic Adenocarcinoma       PL4    Digestive  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapper = pd.read_csv(\n",
    "    input_path+\"Tissue_System_Map.csv\"\n",
    ")\n",
    "mapper.head()\n",
    "\n",
    "# Map the tissue type to system\n",
    "depmap_info = depmap_info.reset_index().merge(\n",
    "    mapper, \n",
    "    how=\"left\", \n",
    "    left_on=\"OncotreeLineage\", \n",
    "    right_on=\"Tissue\"\n",
    ").drop(\n",
    "    \"Tissue\", axis=1\n",
    ").rename(\n",
    "    columns={\"System\": \"System_name\"}\n",
    ")\n",
    "print(\"Number of Cell Line info in DepMap:\", depmap_info.shape[0])\n",
    "depmap_info.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbaf35a8",
   "metadata": {},
   "source": [
    "## Manual Curation of the Metadata"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a6eb22ea-9aad-4792-81db-537b7a5ac878",
   "metadata": {},
   "source": [
    "### Clean and Save Annotations for Manual Editing (Filling Gaps)\n",
    "\n",
    "I save the cell line info table to raw data folder as `manuscript_cell_line_annotations.csv` for manual editing. Which I've already done and created the `manuscript_cell_line_annotations-manual.csv` file. But I am keepnig this here to highlight the process.\n",
    "\n",
    "THe saved file has simpliefied column names and only contains the cell line models that are present in the manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9e5825ce-d817-4e51-825b-7200cfd6ceee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SIDM</th>\n",
       "      <th>Cell_line</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Status</th>\n",
       "      <th>Collection_site</th>\n",
       "      <th>System_name</th>\n",
       "      <th>Tissue_type</th>\n",
       "      <th>Cancer_type</th>\n",
       "      <th>Cancer_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM00772</td>\n",
       "      <td>HCC2218</td>\n",
       "      <td>Female</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Primary</td>\n",
       "      <td>breast</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast Ductal Carcinoma In Situ</td>\n",
       "      <td>Breast Ductal Carcinoma In Situ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00872</td>\n",
       "      <td>HCC1954</td>\n",
       "      <td>Female</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Primary</td>\n",
       "      <td>breast</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast Ductal Carcinoma In Situ</td>\n",
       "      <td>Breast Ductal Carcinoma In Situ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM00875</td>\n",
       "      <td>HCC1806</td>\n",
       "      <td>Female</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Primary</td>\n",
       "      <td>breast</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast Ductal Carcinoma In Situ</td>\n",
       "      <td>Breast Ductal Carcinoma In Situ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDM00879</td>\n",
       "      <td>HCC1500</td>\n",
       "      <td>Female</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Primary</td>\n",
       "      <td>breast</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast Ductal Carcinoma In Situ</td>\n",
       "      <td>Breast Ductal Carcinoma In Situ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDM00954</td>\n",
       "      <td>COLO-824</td>\n",
       "      <td>Female</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Metastatic</td>\n",
       "      <td>pleural_effusion</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Breast Neoplasm, NOS</td>\n",
       "      <td>Breast Neoplasm, NOS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        SIDM Cell_line     Sex    Age      Status   Collection_site  \\\n",
       "0  SIDM00772   HCC2218  Female  Adult     Primary            breast   \n",
       "1  SIDM00872   HCC1954  Female  Adult     Primary            breast   \n",
       "2  SIDM00875   HCC1806  Female  Adult     Primary            breast   \n",
       "3  SIDM00879   HCC1500  Female  Adult     Primary            breast   \n",
       "4  SIDM00954  COLO-824  Female  Adult  Metastatic  pleural_effusion   \n",
       "\n",
       "  System_name Tissue_type                      Cancer_type  \\\n",
       "0      Breast      Breast  Breast Ductal Carcinoma In Situ   \n",
       "1      Breast      Breast  Breast Ductal Carcinoma In Situ   \n",
       "2      Breast      Breast  Breast Ductal Carcinoma In Situ   \n",
       "3      Breast      Breast  Breast Ductal Carcinoma In Situ   \n",
       "4      Breast      Breast             Breast Neoplasm, NOS   \n",
       "\n",
       "                    Cancer_subtype  \n",
       "0  Breast Ductal Carcinoma In Situ  \n",
       "1  Breast Ductal Carcinoma In Situ  \n",
       "2  Breast Ductal Carcinoma In Situ  \n",
       "3  Breast Ductal Carcinoma In Situ  \n",
       "4             Breast Neoplasm, NOS  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean-up and save the data\n",
    "depmap_info.columns = [\n",
    "    \"SIDM\",\n",
    "    \"Age\",\n",
    "    \"Status\",\n",
    "    \"Sex\",\n",
    "    \"Collection_site\",\n",
    "    \"Tissue_type\",\n",
    "    \"Cancer_type\",\n",
    "    \"Cancer_subtype\",\n",
    "    \"Cell_line\",\n",
    "    \"System_name\"\n",
    "]\n",
    "\n",
    "# Order the columns\n",
    "depmap_info = depmap_info[[\n",
    "    \"SIDM\",\n",
    "    \"Cell_line\",\n",
    "    \"Sex\",\n",
    "    \"Age\",\n",
    "    \"Status\",\n",
    "    \"Collection_site\",\n",
    "    \"System_name\",\n",
    "    \"Tissue_type\",\n",
    "    \"Cancer_type\",\n",
    "    \"Cancer_subtype\"\n",
    "]].sort_values([\n",
    "    \"System_name\",\n",
    "    \"Tissue_type\",\n",
    "    \"Cancer_type\",\n",
    "    \"Cancer_subtype\",\n",
    "]).reset_index(drop=True)\n",
    "\n",
    "# Save the data\n",
    "depmap_info.to_csv(\n",
    "    input_path + \"manuscript_cell_line_annotations.csv\",\n",
    "    index=False\n",
    ")\n",
    "\n",
    "# Preview the data\n",
    "depmap_info.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f6c29833-1abc-47d6-a018-d561578e107f",
   "metadata": {},
   "source": [
    "### Manual Annotation\n",
    "\n",
    "In the manual annotation I use multiple sources to fill the missing information in the metadata. I mainly used the [Cellosaurus](https://www.cellosaurus.org/) resource as well as other google to see if any information can be found. Overall I was able to fill the missing information about half of the times. And the ones I wasn't able to fill I left them as Unknown or NaN.\n",
    "\n",
    "Also I've created a column called \"Drop\" while I was doing the manual curation, if the value inside is 1 then I will drop that cell line when I am done with the manual curation."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "873d786c-807b-43c2-a4d0-00eb2f4c4ad8",
   "metadata": {},
   "source": [
    "### Clean the Manual Annotations\n",
    "\n",
    "After opening the manually annotated metadata I clean the data by removing the cell lines that have missing info labeled with Drop = 1. Also I've removed the cell lines with Unknown Sex and Unknown or Fetus Age since I want to focus on Adult and Childhood cancers. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "539f4ea8-e2a0-44b7-8f5c-ce3d42b56c15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data Shape: (950, 11)\n",
      "Data after manually labeled for dropping is done: (814, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SIDM</th>\n",
       "      <th>Cell_line</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Status</th>\n",
       "      <th>Collection_site</th>\n",
       "      <th>System_name</th>\n",
       "      <th>Tissue_type</th>\n",
       "      <th>Cancer_type</th>\n",
       "      <th>Cancer_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM00287</td>\n",
       "      <td>BALL-1</td>\n",
       "      <td>Male</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Peripheral_blood</td>\n",
       "      <td>Hematologic</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>Acute Lymphoblastic Leukemia</td>\n",
       "      <td>Adult B-Lymphoblastic Leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00429</td>\n",
       "      <td>NALM-6</td>\n",
       "      <td>Male</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Peripheral_blood</td>\n",
       "      <td>Hematologic</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>Acute Lymphoblastic Leukemia</td>\n",
       "      <td>Adult B-Lymphoblastic Leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM00813</td>\n",
       "      <td>RPMI-6666</td>\n",
       "      <td>Male</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Peripheral_blood</td>\n",
       "      <td>Hematologic</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>Acute Lymphoblastic Leukemia</td>\n",
       "      <td>Adult B-Lymphoblastic Leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDM01086</td>\n",
       "      <td>RS4-11</td>\n",
       "      <td>Female</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Bone_marrow</td>\n",
       "      <td>Hematologic</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>Acute Lymphoblastic Leukemia</td>\n",
       "      <td>Adult B-Lymphoblastic Leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDM00415</td>\n",
       "      <td>ROS-50</td>\n",
       "      <td>Male</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Peripheral_blood</td>\n",
       "      <td>Hematologic</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>Acute Lymphoblastic Leukemia</td>\n",
       "      <td>Adult B-Lymphoblastic Leukemia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        SIDM  Cell_line     Sex    Age   Status   Collection_site  \\\n",
       "0  SIDM00287     BALL-1    Male  Adult  Primary  Peripheral_blood   \n",
       "1  SIDM00429     NALM-6    Male  Adult  Primary  Peripheral_blood   \n",
       "2  SIDM00813  RPMI-6666    Male  Adult  Primary  Peripheral_blood   \n",
       "3  SIDM01086     RS4-11  Female  Adult  Primary       Bone_marrow   \n",
       "4  SIDM00415     ROS-50    Male  Adult  Primary  Peripheral_blood   \n",
       "\n",
       "   System_name Tissue_type                   Cancer_type  \\\n",
       "0  Hematologic    Lymphoid  Acute Lymphoblastic Leukemia   \n",
       "1  Hematologic    Lymphoid  Acute Lymphoblastic Leukemia   \n",
       "2  Hematologic    Lymphoid  Acute Lymphoblastic Leukemia   \n",
       "3  Hematologic    Lymphoid  Acute Lymphoblastic Leukemia   \n",
       "4  Hematologic    Lymphoid  Acute Lymphoblastic Leukemia   \n",
       "\n",
       "                   Cancer_subtype  \n",
       "0  Adult B-Lymphoblastic Leukemia  \n",
       "1  Adult B-Lymphoblastic Leukemia  \n",
       "2  Adult B-Lymphoblastic Leukemia  \n",
       "3  Adult B-Lymphoblastic Leukemia  \n",
       "4  Adult B-Lymphoblastic Leukemia  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_info_cur = pd.read_csv(input_path+\"manuscript_cell_line_annotations-manual.csv\")\n",
    "print(\"Data Shape:\", cell_info_cur.shape)\n",
    "# Remove the cell lines with drop = 1\n",
    "cell_info_cur = cell_info_cur[cell_info_cur[\"Drop\"]!=1]\n",
    "cell_info_cur = cell_info_cur.drop(columns=\"Drop\").reset_index(drop=True)\n",
    "\n",
    "# Remove cell lines (Sex = Unknown & Age = [Unknown, Fetus])\n",
    "cell_info_cur = cell_info_cur[\n",
    "    (cell_info_cur[\"Sex\"]!=\"Unknown\") & \n",
    "    (~cell_info_cur[\"Age\"].isin([\"Unknown\", \"Fetus\"]))\n",
    "]\n",
    "\n",
    "cell_info_cur[\"Age\"] = cell_info_cur[\"Age\"].replace(\"Pediatric\", \"Child\")\n",
    "\n",
    "cell_info_cur.sort_values([\n",
    "    \"System_name\",\n",
    "    \"Tissue_type\",\n",
    "    \"Cancer_type\",\n",
    "    \"Cancer_subtype\",\n",
    "    \"Status\",\n",
    "    \"Age\",\n",
    "]).reset_index(drop=True)\n",
    "\n",
    "print(\"Data after manually labeled for dropping is done:\", cell_info_cur.shape)\n",
    "cell_info_cur.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f63d2d6a",
   "metadata": {},
   "source": [
    "After the dropping some of the cell line based on the criterias above from 950 cell lines I am left with 814 cell lines."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a9ec778e-3089-4f7a-894e-09eeca2ac5dc",
   "metadata": {},
   "source": [
    "## Replica and Cell Line Metadata is Ready\n",
    "\n",
    "Since we dropped some cell lines in the manual curation step, I will up date the replica metadata as well and ensure they are in sync with the cell line metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "117ce6a6-34e9-4d25-b124-9c8149b23970",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Replicate data before: 5800\n",
      "Replicate data after: 4978\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Automatic_MS_filename</th>\n",
       "      <th>ColumnName</th>\n",
       "      <th>UniqueSample</th>\n",
       "      <th>VarianceID</th>\n",
       "      <th>Cell_line</th>\n",
       "      <th>Replicate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SIDM00183</th>\n",
       "      <td>200117_b1-8-t1-1_00q2c_00qi2_m01_s_1</td>\n",
       "      <td>M01|P06|SIDM00183|1</td>\n",
       "      <td>M01|P06|SIDM00183</td>\n",
       "      <td>M01|P06</td>\n",
       "      <td>LB647-SCLC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00183</th>\n",
       "      <td>200130_b8-1-t2-1_00q59_00rp7_m01_s_1</td>\n",
       "      <td>M01|P06|SIDM00183|2</td>\n",
       "      <td>M01|P06|SIDM00183</td>\n",
       "      <td>M01|P06</td>\n",
       "      <td>LB647-SCLC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00183</th>\n",
       "      <td>200118_b1-8-t2-1_00q2c_00qim_m03_s_1</td>\n",
       "      <td>M03|P06|SIDM00183|1</td>\n",
       "      <td>M03|P06|SIDM00183</td>\n",
       "      <td>M03|P06</td>\n",
       "      <td>LB647-SCLC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00183</th>\n",
       "      <td>200121_b3-8-t2-1_00q38_00qko_m03_s_1</td>\n",
       "      <td>M03|P06|SIDM00183|2</td>\n",
       "      <td>M03|P06|SIDM00183</td>\n",
       "      <td>M03|P06</td>\n",
       "      <td>LB647-SCLC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00183</th>\n",
       "      <td>200129_b8-1-t1-1_00q59_00roe_m05_s_1</td>\n",
       "      <td>M05|P06|SIDM00183|1</td>\n",
       "      <td>M05|P06|SIDM00183</td>\n",
       "      <td>M05|P06</td>\n",
       "      <td>LB647-SCLC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Automatic_MS_filename           ColumnName  \\\n",
       "SIDM                                                                   \n",
       "SIDM00183  200117_b1-8-t1-1_00q2c_00qi2_m01_s_1  M01|P06|SIDM00183|1   \n",
       "SIDM00183  200130_b8-1-t2-1_00q59_00rp7_m01_s_1  M01|P06|SIDM00183|2   \n",
       "SIDM00183  200118_b1-8-t2-1_00q2c_00qim_m03_s_1  M03|P06|SIDM00183|1   \n",
       "SIDM00183  200121_b3-8-t2-1_00q38_00qko_m03_s_1  M03|P06|SIDM00183|2   \n",
       "SIDM00183  200129_b8-1-t1-1_00q59_00roe_m05_s_1  M05|P06|SIDM00183|1   \n",
       "\n",
       "                UniqueSample VarianceID   Cell_line Replicate  \n",
       "SIDM                                                           \n",
       "SIDM00183  M01|P06|SIDM00183    M01|P06  LB647-SCLC         1  \n",
       "SIDM00183  M01|P06|SIDM00183    M01|P06  LB647-SCLC         2  \n",
       "SIDM00183  M03|P06|SIDM00183    M03|P06  LB647-SCLC         1  \n",
       "SIDM00183  M03|P06|SIDM00183    M03|P06  LB647-SCLC         2  \n",
       "SIDM00183  M05|P06|SIDM00183    M05|P06  LB647-SCLC         1  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Replicate data before:\", replica_meta.shape[0])\n",
    "replica_meta = replica_meta.set_index(\"SIDM\").loc[list(set(cell_info_cur[\"SIDM\"]))]\n",
    "print(\"Replicate data after:\", replica_meta.shape[0])\n",
    "replica_meta.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "715ddf9f-ea9e-472a-af99-ce2437beedb0",
   "metadata": {},
   "source": [
    "### Save the Final Data\n",
    "\n",
    "The cleaned replicate metadata and expanded cell line metadata is saved in the `data/raw/` folder as `cleaned_replica_metadata.csv` and `final_cell_line_annotation.csv` respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bbc1b21e-59df-444f-a005-dd840ed2cca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_info_cur.to_csv(input_path+\"final_cell_line_annotation.csv\", index=False)\n",
    "replica_meta.to_csv(input_path+\"cleaned_replica_metadata.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f57b7e6d",
   "metadata": {},
   "source": [
    "# Helper Data Tables\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49e7a574",
   "metadata": {},
   "source": [
    "## UniProt Protein Information\n",
    "\n",
    "Having access to the expanded protein information is useful when reporting the results with more than just protien accession ids. I've downloaded the human \"reviewed proteome without isoforms\" fasta from [Uniprot](https://www.uniprot.org/) accessed on 2023-11-01 and saved it in the `./data/raw/` folder as `HUMAN_REVIEWED_2023_11_01.fasta`.\n",
    "\n",
    "The using the `fasta_to_df` function from `utils.py` script I've parsed the fasta file and created a dataframe: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d876890e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data Shape: (20265, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry</th>\n",
       "      <th>entryName</th>\n",
       "      <th>geneName</th>\n",
       "      <th>proteinDescription</th>\n",
       "      <th>sequenceLength</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A024R1R8</td>\n",
       "      <td>TMA7B_HUMAN</td>\n",
       "      <td>TMA7B</td>\n",
       "      <td>Translation machinery-associated protein 7B</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A024RBG1</td>\n",
       "      <td>NUD4B_HUMAN</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>Diphosphoinositol polyphosphate phosphohydrola...</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A075B6H7</td>\n",
       "      <td>KV37_HUMAN</td>\n",
       "      <td>IGKV3-7</td>\n",
       "      <td>Probable non-functional immunoglobulin kappa v...</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A075B6H8</td>\n",
       "      <td>KVD42_HUMAN</td>\n",
       "      <td>IGKV1D-42</td>\n",
       "      <td>Probable non-functional immunoglobulin kappa v...</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A075B6H9</td>\n",
       "      <td>LV469_HUMAN</td>\n",
       "      <td>IGLV4-69</td>\n",
       "      <td>Immunoglobulin lambda variable 4-69</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        entry    entryName   geneName  \\\n",
       "0  A0A024R1R8  TMA7B_HUMAN      TMA7B   \n",
       "1  A0A024RBG1  NUD4B_HUMAN     NUDT4B   \n",
       "2  A0A075B6H7   KV37_HUMAN    IGKV3-7   \n",
       "3  A0A075B6H8  KVD42_HUMAN  IGKV1D-42   \n",
       "4  A0A075B6H9  LV469_HUMAN   IGLV4-69   \n",
       "\n",
       "                                  proteinDescription  sequenceLength  \n",
       "0        Translation machinery-associated protein 7B              64  \n",
       "1  Diphosphoinositol polyphosphate phosphohydrola...             181  \n",
       "2  Probable non-functional immunoglobulin kappa v...             116  \n",
       "3  Probable non-functional immunoglobulin kappa v...             117  \n",
       "4                Immunoglobulin lambda variable 4-69             119  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_data = utils.fasta_to_df(\n",
    "    f\"{input_path}HUMAN_REVIEWED_2023_11_01.fasta\",\n",
    "    fasta_ID=\"\" # This is not needed for this usecase\n",
    ").drop(\n",
    "    columns=[\n",
    "        \"fastaId\", \n",
    "        \"reviewStatus\", \n",
    "        \"isoformStatus\", \n",
    "        \"molecularWeight_kDa\",\n",
    "        \"sequence\",\n",
    "    ]\n",
    ")\n",
    "# # Save the data as feather\n",
    "# feather.write_dataframe(\n",
    "#     fasta_data, \n",
    "#     input_path+\"UniProtIDMapping.feather\"\n",
    "# )\n",
    "fasta_data.to_csv(\n",
    "    input_path+\"UniProtIDMapping.csv\", \n",
    "    index=False\n",
    ")\n",
    "\n",
    "print(\"Data Shape:\", fasta_data.shape)\n",
    "fasta_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8237d655",
   "metadata": {},
   "source": [
    "Resulting dataframe is saved in the `./data/raw/` folder as `UniProtIDMapping.csv`, I opted to save it as csv since it is easier to read and edit."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c291fc9d",
   "metadata": {},
   "source": [
    "## Reactome Pathway Hierarchy\n",
    "\n",
    "The reactome pathway hierarchy is not directly available when I use enrichment analysis results from `g:Profiler` that's why I want to createa a table that enables a quick lookup of the pathway hierarchy as well as a mechanism to group pathways into highest-level parent pathways.\n",
    "\n",
    "The reactome pathway hierarchy is downloaded from the [Reactome](https://reactome.org/download-data) accessed on 2023-11-01 and saved in the `./data/raw/` folder as `ReactomePathways.txt` and `ReactomePathwaysRelation.txt`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1cf83b15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Data Shape: (22084, 3)\n",
      "Number of pathways per organism\n",
      " Homo sapiens -> 2673\n",
      " Mus musculus -> 1737\n",
      " Gallus gallus -> 1736\n",
      " Rattus norvegicus -> 1724\n",
      " Bos taurus -> 1718\n",
      " Sus scrofa -> 1709\n",
      " Canis familiaris -> 1690\n",
      " Drosophila melanogaster -> 1511\n",
      " Xenopus tropicalis -> 1502\n",
      " Danio rerio -> 1353\n",
      " Caenorhabditis elegans -> 1350\n",
      " Dictyostelium discoideum -> 1024\n",
      " Schizosaccharomyces pombe -> 864\n",
      " Saccharomyces cerevisiae -> 850\n",
      " Plasmodium falciparum -> 630\n",
      " Mycobacterium tuberculosis -> 13\n",
      "\n",
      "Data Shape after: (2673, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PathwayID</th>\n",
       "      <th>PathwayName</th>\n",
       "      <th>Organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R-HSA-164843</td>\n",
       "      <td>2-LTR circle formation</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R-HSA-73843</td>\n",
       "      <td>5-Phosphoribose 1-diphosphate biosynthesis</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R-HSA-1971475</td>\n",
       "      <td>A tetrasaccharide linker sequence is required ...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R-HSA-5619084</td>\n",
       "      <td>ABC transporter disorders</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R-HSA-1369062</td>\n",
       "      <td>ABC transporters in lipid homeostasis</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PathwayID                                        PathwayName  \\\n",
       "0   R-HSA-164843                             2-LTR circle formation   \n",
       "1    R-HSA-73843         5-Phosphoribose 1-diphosphate biosynthesis   \n",
       "2  R-HSA-1971475  A tetrasaccharide linker sequence is required ...   \n",
       "3  R-HSA-5619084                          ABC transporter disorders   \n",
       "4  R-HSA-1369062              ABC transporters in lipid homeostasis   \n",
       "\n",
       "       Organism  \n",
       "0  Homo sapiens  \n",
       "1  Homo sapiens  \n",
       "2  Homo sapiens  \n",
       "3  Homo sapiens  \n",
       "4  Homo sapiens  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reactome_pathways = pd.read_csv(\n",
    "    f\"{input_path}ReactomePathways.txt\", sep=\"\\t\", header=None\n",
    ")\n",
    "print()\n",
    "reactome_pathways.columns = [\"PathwayID\", \"PathwayName\", \"Organism\"]\n",
    "print(\"Data Shape:\", reactome_pathways.shape)\n",
    "utils.print_series(\n",
    "    reactome_pathways[\"Organism\"].value_counts(),\n",
    "    header=\"Number of pathways per organism\"\n",
    ")\n",
    "print()\n",
    "# Subset for only human pathways\n",
    "reactome_pathways = reactome_pathways[\n",
    "    reactome_pathways[\"Organism\"] == \"Homo sapiens\"\n",
    "].reset_index(drop=True)\n",
    "print(\"Data Shape after:\", reactome_pathways.shape)\n",
    "reactome_pathways.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a3ad379",
   "metadata": {},
   "source": [
    "For the purposes of this analysis I am only interested in the human pathways, so I filter the data to only contain the human pathways. Then I open the hierarcy data and create a dictionary that contains the pathway hierarchy. and I make sure to only get the hierarcy for the human pathways.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c90a284c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data Shape: (2665, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent</th>\n",
       "      <th>Child</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R-HSA-109581</td>\n",
       "      <td>R-HSA-109606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R-HSA-109581</td>\n",
       "      <td>R-HSA-169911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R-HSA-109581</td>\n",
       "      <td>R-HSA-5357769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R-HSA-109581</td>\n",
       "      <td>R-HSA-75153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R-HSA-109582</td>\n",
       "      <td>R-HSA-140877</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Parent          Child\n",
       "0  R-HSA-109581   R-HSA-109606\n",
       "1  R-HSA-109581   R-HSA-169911\n",
       "2  R-HSA-109581  R-HSA-5357769\n",
       "3  R-HSA-109581    R-HSA-75153\n",
       "4  R-HSA-109582   R-HSA-140877"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paired_hierarchy = pd.read_csv(\n",
    "    f\"{input_path}ReactomePathwaysRelation.txt\",\n",
    "    sep=\"\\t\", \n",
    "    header=None\n",
    ")\n",
    "paired_hierarchy.columns = [\"Parent\", \"Child\"]\n",
    "\n",
    "# Subset only Human pathways in paired_hierarcy\n",
    "paired_hierarchy = paired_hierarchy[\n",
    "    (paired_hierarchy[\"Parent\"].isin(reactome_pathways[\"PathwayID\"])) | \n",
    "    (paired_hierarchy[\"Child\"].isin(reactome_pathways[\"PathwayID\"]))\n",
    "].reset_index(drop=True)\n",
    "\n",
    "print(\"Data Shape:\", paired_hierarchy.shape)\n",
    "paired_hierarchy.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6736359b",
   "metadata": {},
   "source": [
    "To facilitate the lookup of the pathway hierarchy I merge the parent pathway information to the main reactome pathway data table. As an additional information I create a column called \"NumChildren\" which will indicate how many iterative children pathway are there for the given pathway. If the pathway is at the very bottom of the hierarchy then the value will be 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "30e7ff5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of mainParentIDs: 29\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PathwayID</th>\n",
       "      <th>PathwayName</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Parent</th>\n",
       "      <th>NumChildren</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R-HSA-164843</td>\n",
       "      <td>2-LTR circle formation</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>R-HSA-162592</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R-HSA-73843</td>\n",
       "      <td>5-Phosphoribose 1-diphosphate biosynthesis</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>R-HSA-71336</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R-HSA-1971475</td>\n",
       "      <td>A tetrasaccharide linker sequence is required ...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>R-HSA-1793185</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R-HSA-5619084</td>\n",
       "      <td>ABC transporter disorders</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>R-HSA-5619115</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R-HSA-1369062</td>\n",
       "      <td>ABC transporters in lipid homeostasis</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>R-HSA-382556</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PathwayID                                        PathwayName  \\\n",
       "0   R-HSA-164843                             2-LTR circle formation   \n",
       "1    R-HSA-73843         5-Phosphoribose 1-diphosphate biosynthesis   \n",
       "2  R-HSA-1971475  A tetrasaccharide linker sequence is required ...   \n",
       "3  R-HSA-5619084                          ABC transporter disorders   \n",
       "4  R-HSA-1369062              ABC transporters in lipid homeostasis   \n",
       "\n",
       "       Organism         Parent  NumChildren  \n",
       "0  Homo sapiens   R-HSA-162592            0  \n",
       "1  Homo sapiens    R-HSA-71336            0  \n",
       "2  Homo sapiens  R-HSA-1793185            0  \n",
       "3  Homo sapiens  R-HSA-5619115           15  \n",
       "4  Homo sapiens   R-HSA-382556            0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def count_children(term, term_dict):\n",
    "    # Base case: The term is not in the dictionary (it has no children)\n",
    "    if term not in term_dict:\n",
    "        return 0\n",
    "\n",
    "    # Recursive case: The term is in the dictionary\n",
    "    else:\n",
    "        children = term_dict[term]\n",
    "        count = len(children)  # Count the direct children\n",
    "        for child in children:\n",
    "            count += count_children(child, term_dict)  # Add the count of the nested children\n",
    "        return count\n",
    "\n",
    "# Identify the highest hierarchy Parents\n",
    "# Find Parent ids that are not in Child\n",
    "mainParent_IDs = list(\n",
    "    set(paired_hierarchy[\"Parent\"].tolist()) - \n",
    "    set(paired_hierarchy[\"Child\"].tolist())\n",
    ")\n",
    "print(\"The number of mainParentIDs:\", len(mainParent_IDs))\n",
    "\n",
    "# Create Parent: All Children Dictionary\n",
    "term_dict = paired_hierarchy.groupby('Parent')['Child'].apply(list).to_dict()\n",
    "term_name_dict = reactome_pathways.set_index(\"PathwayID\")[\"PathwayName\"].to_dict()\n",
    "# Add the Parent column to the reactome_pathways dataframe\n",
    "reactome_pathways[\"Parent\"] = reactome_pathways[\"PathwayID\"].map(\n",
    "    paired_hierarchy.set_index(\"Child\")[\"Parent\"].to_dict()\n",
    ")\n",
    "# Add the number of children to the reactome_pathways dataframe\n",
    "reactome_pathways[\"NumChildren\"] = reactome_pathways[\"PathwayID\"].apply(\n",
    "    lambda x: count_children(x, term_dict)\n",
    ")\n",
    "\n",
    "# Save the data\n",
    "reactome_pathways[[\n",
    "    \"PathwayID\", \"PathwayName\", \"Parent\", \"NumChildren\"\n",
    "]].to_csv(\n",
    "    f\"{input_path}ReactomeDetailed.txt\", \n",
    "    sep=\"\\t\", \n",
    "    index=False\n",
    ")\n",
    "\n",
    "reactome_pathways.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b552a0c",
   "metadata": {},
   "source": [
    "The resulting dataframe is saved in the `./data/raw/` folder as `ReactomeDetailed.txt`, I opted to save it as a txt file since the this data won't be needed to be opened frequently. \n",
    "\n",
    "---\n",
    "\n",
    "Note: Not related to the creating of the data, but I've also created a function called list_children which neatly prints the hierarchy of the given pathway. Here are some examples of how it looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "216cdeb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R-HSA-75072 - mRNA Editing\n",
      "\n",
      "\tmRNA Editing: A to I Conversion\n",
      "mRNA Editing: C to U Conversion\n",
      "\tFormation of the Editosome\n",
      "\tC6 deamination of adenosine\n",
      "Formation of editosomes by ADAR proteins\n"
     ]
    }
   ],
   "source": [
    "def list_children(term, term_dict, term_name_dict):\n",
    "    \"\"\"Create a neat new line and tab listed print of the children of a term\"\"\"\n",
    "    # Base case: The term is not in the dictionary (it has no children)\n",
    "    if term not in term_dict:\n",
    "        return \"\"\n",
    "\n",
    "    # Recursive case: The term is in the dictionary\n",
    "    else:\n",
    "        children = term_dict[term]\n",
    "        children_names = [term_name_dict[child] for child in children]\n",
    "        children_names.sort()\n",
    "        children_names = \"\\n\".join(children_names)\n",
    "        children_names = \"\\n\\t\" + children_names\n",
    "        for child in children:\n",
    "            children_names += list_children(child, term_dict, term_name_dict)\n",
    "        return children_names\n",
    "    \n",
    "term = \"R-HSA-75072\" \n",
    "print(term ,\"-\" , term_name_dict[term])\n",
    "print(list_children(term, term_dict, term_name_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "228ce742",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R-HSA-72306 - tRNA processing\n",
      "\n",
      "\ttRNA modification in the mitochondrion\n",
      "tRNA modification in the nucleus and cytosol\n",
      "tRNA processing in the mitochondrion\n",
      "tRNA processing in the nucleus\n",
      "\tSynthesis of wybutosine at G37 of tRNA(Phe)\n"
     ]
    }
   ],
   "source": [
    "term = \"R-HSA-72306\"\n",
    "print(term ,\"-\" , term_name_dict[term])\n",
    "print(list_children(term, term_dict, term_name_dict))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "364008f7",
   "metadata": {},
   "source": [
    "# Conclusion\n",
    "\n",
    "This notebook is the first step in the analysis of the cancer cell line data. In this notebook I've prepared the a comprehensive metadata for the cell lines that will be used in the analysis and other useful helper tables for reactome hierarchies and uniprot protein information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "da48b2b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook Execution Time: 00h:00m:02s\n"
     ]
    }
   ],
   "source": [
    "print(\"Notebook Execution Time:\", utils.prettyTimer(utils.getTime() - startTime))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
